### ISSUE TITLE
[Feature Request] Supporting new `native CSS nesting`

### ISSUE THREAD

**Issue by hoangnhan2ka3** [soranoo/next-css-obfuscator#20]
**Checklist**
1. [x] Read all documentation
2. [x] No related request
3. [x] Meaningful issue title

**Is your feature request related to a problem? Please describe.**
Maybe not, just a `feature request`.

**Describe the solution you'd like**
Supporting obfuscate new `native CSS nesting`.

**Describe alternatives you've considered**
`1.` first, `2.` later.

**Additional context**

**👀Version `v2.2.7`**

> _Note: sorry for the inconvenience, `developer.mozilla.org` doesn't seem to automatically scroll to the location of `ref` links in this article (at least for me), you can scroll until If you see any text that is `highlighted` to see it._

&#32;
&#32;
### **1. With `CSS nesting plugin *before* Tailwind` enabled.**
✅ `next-css-obfuscator` work with these plugin enabled! (but there's still a bug which will not replaced the obfuscated class)

### **2. With `CSS nesting plugin *before* Tailwind` disabled.**
⛔ `next-css-obfuscator` does not work when not enable these plugin! (I can't capture the log cuz it's too long)
&#32;
&#32;

So, what is that plugin?

🐽 1. When you using `CSS Nesting` in NextJs with TailwindCSS, you will receive a notification like this in the `log` terminal:

```terminal
(82:2) Nested CSS was detected, but CSS nesting has not been configured correctly.
Please enable a CSS nesting plugin *before* Tailwind in your configuration.
See how here: https://tailwindcss.com/docs/using-with-preprocessors#nesting
```

And in the instruction link it provides, follow it, and you will get results like this in `postcss.config.js`:

![image](https://github.com/soranoo/next-css-obfuscator/assets/147973044/685ffb98-5ccc-4695-ad87-39935d61434f)

Now `CSS Nesting` in your `global.css` file will behave much like `.sass` after `npm run build`. That means that even though existing CSS supports `native CSS nesting`, TailwindCSS will still convert it to `normal CSS`.

🐽 2. But if you do nothing, just ignore it, it's still works (cuz it became native now), but everytime you `save` the file, TailwindCSS will yell the above `warning`

> It still works but is not supported by default by TailwindCSS, NextJs...

Before going to the `Big sections`, let's explain a bit.

### **✨ Main point:**

> CSS nesting is different from CSS preprocessors such as [Sass](https://sass-lang.com/) in that it is parsed by the browser rather than being pre-compiled by a CSS preprocessor. [(ref)](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_nesting/Using_CSS_nesting#:~:text=CSS%20nesting%20is%20different%20from%20CSS%20preprocessors%20such%20as%20Sass%20in%20that%20it%20is%20parsed%20by%20the%20browser%20rather%20than%20being%20pre%2Dcompiled%20by%20a%20CSS%20preprocessor.)

Example TSX:

```tsx
<div className="note">
  <p>
    <b className="note-title">Note:</b> This demo website is just <u id="note-exp">experimental</u>, so if you see something wrong <i>(like error layout or typography)</i>, please do a &quot;hard reload&quot; (<b>Ctrl + Shift + R</b> or <b>Cmd + Shift + R</b>)
  </p>
</div>
```

👉 In fact, CSS already supports `nesting` natively in the form of `nesting selector`: `&` like below [(ref)](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_nesting/Using_CSS_nesting#:~:text=Play-,With%20nesting%20selector,-HTML):

```css
/* global.css */

.note {
  display: flex;
  align-items: center;
  transition: opacity 0.2s ease-out;
  transition-delay: 1.2s;
  opacity: 0.2;
  cursor: help;

  /* class */
  & .note-title {
    letter-spacing: 1rem;
  }

  /* id */
  & #note-exp {
    opacity: 0.2;
  }

  /* pseudo-class */
  &:hover {
    opacity: 1;
    transition-delay: 0s;
  }
  /*! `&:hover` not `& :hover` <- WITH blank no work */

  /* tag */
  & p {
    font-size: 0.875rem;
  }
  /*! `& p` not `&p` <- NO blank no work */
}
```

👉 But recently (I don't know when), CSS has supported `without nesting selector` [(ref)](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_nesting/Using_CSS_nesting#:~:text=the%20second%20example.-,Without%20nesting%20selector,-HTML):

```css
/* global.css */

.note {
  display: flex;
  align-items: center;
  transition: opacity 0.2s ease-out;
  transition-delay: 1.2s;
  opacity: 0.2;
  cursor: help;

  /* class */
  .note-title {
    letter-spacing: 1rem;
  }

  /* id */
  #note-exp {
    opacity: 0.2;
  }

  /* pseudo-class */
  &:hover {
    opacity: 1;
    transition-delay: 0s;
  }
  /*! `&:hover` not `& :hover` <- WITH blank no work */

  /* tag */
  p {
    font-size: 0.875rem;
  }
}
```

_Hmmm, except something like `:hover`, `::before`, `::after` maybe, I tried and it didn't work without `&` in front. (I didn't try with `:is`cuz I can't think of a suitable case)_

👉 On the other hand, there exists something called `CSS Combinators`: `+` [(ref)](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_nesting/Using_CSS_nesting#see_also:~:text=Combinators-,CSS%20Combinators,-can%20also%20be)

> The next-sibling combinator (+) separates two selectors and matches the second element only if it immediately **FOLLOWS THE FIRST ELEMENT**, and both are children of the same parent element. [(ref)](https://developer.mozilla.org/en-US/docs/Web/CSS/Next-sibling_combinator#:~:text=The%20next%2Dsibling%20combinator%20(%2B)%20separates%20two%20selectors%20and%20matches%20the%20second%20element%20only%20if%20it%20immediately%20follows%20the%20first%20element%2C%20and%20both%20are%20children%20of%20the%20same%20parent%20element.)

```css
h2 {
  color: tomato;
  + p {
    color: white;
    background-color: black;
  }
}
```
Okay, maybe enough...

&#32;
&#32;

### **🚀 1. With `CSS nesting plugin *before* Tailwind` ENABLED.**

![image](https://github.com/soranoo/next-css-obfuscator/assets/147973044/685ffb98-5ccc-4695-ad87-39935d61434f)

**IMPORTANT: it appears that `TailwindCSS` automatically converts these `native CSS nesting` to `normal CSS` (without nesting) via `PostCSS`, even though it doesn't need to be done to run normally** 
_(as evidenced by `run dev`, or disable `next-css-obfuscator` plugin and `run build + run start`)._

&#32;
> `next-css-obfuscator` plugin ENABLED

Obfuscated ✅:

`conversion.json`

<img width="470" alt="Screenshot 2024-02-15 110435" src="https://github.com/soranoo/next-css-obfuscator/assets/147973044/22dbabb6-972a-4433-94db-d90590195bab">

&#32;
Not replaced ⛔:

`output .css file` with `next-css-obfuscator` plugin `enable`:

<img width="354" alt="Screenshot 2024-02-15 110141" src="https://github.com/soranoo/next-css-obfuscator/assets/147973044/097df5f7-37f0-4758-94ae-0d52afbf49b6">

_Please ignore the red comment, it seems to be normal behavior in `production`_

&#32;
> `next-css-obfuscator` plugin DISABLED

`output .css file` with `next-css-obfuscator` plugin `disable`:

<img width="322" alt="Screenshot 2024-02-15 110924" src="https://github.com/soranoo/next-css-obfuscator/assets/147973044/2124373e-89f5-4b98-a7bd-449762e0d1fe">

### **🚀 2. With `CSS nesting plugin *before* Tailwind` DISABLED.**

`next-css-obfuscator` plugin simply doesn't works when `enabled`.

But when disable `next-css-obfuscator`:

This is the `output .css file`:

<img width="313" alt="Screenshot 2024-02-15 125038" src="https://github.com/soranoo/next-css-obfuscator/assets/147973044/ec95bcab-2de6-4846-917b-7c4d7fb2790b">

Which truly a `native CSS nesting`, but when in `production`, it's still yell:

```terminal
(82:2) Nested CSS was detected, but CSS nesting has not been configured correctly.
Please enable a CSS nesting plugin *before* Tailwind in your configuration.
See how here: https://tailwindcss.com/docs/using-with-preprocessors#nesting
```

However, that's just a `warning` not an `error`, the build process still successfully!

### **In conclusion, I think for now you should only focus on fixing part `1.`, as for part `2.`, wait until TW and NextJs support `native CSS nesting` then you can fix it later**

**Comment by hoangnhan2ka3** [comment]
TOO LONG MUST READ !!! 🤣🤣

**Comment by hoangnhan2ka3** [comment]
Not related, but look at your hard work @soranoo , so satisfying 🚀

<img width="750" alt="Screenshot 2024-02-15 151340" src="https://github.com/soranoo/next-css-obfuscator/assets/147973044/2fbc9bf9-c658-4fd2-b0d6-fd906befbcff">

**Comment by soranoo** [comment]
Any sample CSS inside the `.next` folder can be provided? I guess it's just a bug in recursion.

**Comment by hoangnhan2ka3** [comment]
yaeh maybe `1.` is just a bug, but not `2.`

However, u mean this?

[430633817468dcce.css.txt](https://github.com/soranoo/next-css-obfuscator/files/14296733/430633817468dcce.css.txt)

Just delete `.txt` ext or copy it manually.

**Comment by soranoo** [comment]
No, I need the stripped CSS file

**Comment by hoangnhan2ka3** [comment]
U mean the minified one?? but no obfuscated

and would u like me to turn the `CSS nesting plugin` off?

**Comment by hoangnhan2ka3** [comment]
I'll give u two

Obfuscation disable + CSS nesting plugin enable:

[1c37b5e777b9906b.css.txt](https://github.com/soranoo/next-css-obfuscator/files/14297888/1c37b5e777b9906b.css.txt)

Obfuscation disable + CSS nesting plugin disable:

[4c03261ae211d960.css.txt](https://github.com/soranoo/next-css-obfuscator/files/14297909/4c03261ae211d960.css.txt)

**Comment by hoangnhan2ka3** [comment]
Btw, in your `readme.md`:

Line 3: `start -> starts`

Line 41: `sulotion -> solution`

Line 93: `is issue -> is the issue`

Line 95: `change -> changing`

Line 209, 228, 377: `convertion -> conversion`

Line 209: `and obfuscation -> and the obfuscation`

Line 212: `recommanded -> recommended`

Line 313: `still at the early stage -> still in the early stages`

Line 319: `and strongly -> and is strongly`

Line 369: `found the package not work as expected after updated -> find the package does not work as expected after being updated`

Line 414: `may referenced -> may be referenced`

Line 428: `that related to class -> that is related to the class`

Line 431: `component under obfuscation -> component is under the obfuscation`

Line 457: `component -> components` and `children components -> "children" components` (although the right one is `child`, but `children` maybe a `pattern word` in our brain so u can just put it in `" "` :)))

Waiting for the next patch 😶‍🌫️

**Comment by soranoo** [comment]
Please open a new issue if you are going to mention ANYthing not relevant to the current issue!!

**Comment by hoangnhan2ka3** [comment]
Since I'm not a `plugin maker`, I don't understand what it means to have multiple issues in a project :))), but if you want I will, chill~

**Comment by soranoo** [comment]
An issue is something like a topic. It is not relevant whether you are a package creator, It is important to the open-source community!!!

**Comment by hoangnhan2ka3** [comment]
Okay, so sorry, I will do it from now...

What should I choose for that small grammar fix

![image](https://github.com/soranoo/next-css-obfuscator/assets/147973044/3f7b70db-fa6a-44a1-8ab2-553974b70b98)

I think it will be a `feature request`, but do I have the right to remove your entire template and just leave a small suggestion like above?

**Comment by soranoo** [comment]
Checkout #21

**Comment by hoangnhan2ka3** [comment]
Okay...

**Comment by soranoo** [comment]
Case 1: Can't reproduce
Case 2: I'm not going to support nesting in the current step. It would overthrow the foundation of version 2. I will consider implementing it if there is significant demand in the future.

**Comment by hoangnhan2ka3** [comment]
To reproduce:
```js
// postcss.config.js

module.exports = {
  plugins: {
    "postcss-import": {}, // add this
    "tailwindcss/nesting": {}, // and this
    tailwindcss: {},
    autoprefixer: {},
  },
};
```

```cjs
// next-css-obfuscator.config.cjs

module.exports = {
  enable: true,
  mode: "random",
  buildFolderPath: ".next",
  classConversionJsonFolderPath: "./css-obfuscator",
  refreshClassConversionJson: false,

  classLength: 6,
  classPrefix: "n",
  classSuffix: "",

  classIgnore: ["dark", "light", "no-transition", "opa-hidden", "big", "small", /__.*/, /os-*/],

  allowExtensions: [".jsx", ".tsx", ".js", ".ts", ".html", ".rsc"],

  contentIgnoreRegexes: [/\.jsxs\)\("\w+"/g],

  whiteListedFolderPaths: [],

  blackListedFolderPaths: ["./.next/cache"],

  enableMarkers: false,
  markers: ["obf"],
  removeMarkersAfterObfuscated: true,

  removeOriginalCss: true,

  generatorSeed: "84817818898",

  //! Experimental feature (Alpha)
  enableJsAst: false,

  logLevel: "info",
};

```

```tsx
{/* page.tsx */}

...

<div className="note">
  <p>
    <b className="note-title">Note:</b> This demo website is just <u id="note-exp">experimental</u>, so if you see something wrong <i>(like error layout or typography)</i>, please do a &quot;hard reload&quot; (<b>Ctrl + Shift + R</b> or <b>Cmd + Shift + R</b>)
  </p>
</div>

...
```

```css
/* global.css */

@tailwind base;
@tailwind components;
@tailwind utilities;

.note {
  display: flex;
  align-items: center;
  transition: opacity 0.2s ease-out;
  transition-delay: 1.2s;
  opacity: 0.2;
  cursor: help;

  /* class */
  .note-title {
    letter-spacing: 1rem;
  }

  /* id */
  #note-exp {
    opacity: 0.2;
  }

  /* pseudo-class */
  &:hover {
    opacity: 1;
    transition-delay: 0s;
  }
  /*! `&:hover` not `& :hover` <- WITH blank no work */

  /* tag */
  p {
    font-size: 0.875rem;
  }
}
```

=> `yarn build`

Result:

Obfuscated ✅:

`conversion.json`

<img width="470" alt="Screenshot 2024-02-15 110435" src="https://github.com/soranoo/next-css-obfuscator/assets/147973044/22dbabb6-972a-4433-94db-d90590195bab">

&#32;
Not replaced in `output css file` ⛔:

<img width="354" alt="Screenshot 2024-02-15 110141" src="https://github.com/soranoo/next-css-obfuscator/assets/147973044/097df5f7-37f0-4758-94ae-0d52afbf49b6">

=> `yarn start`

Just replaced in `output html file` ✅:

<img width="300" alt="Screenshot 2024-02-16 092353" src="https://github.com/soranoo/next-css-obfuscator/assets/147973044/ea7b3271-a990-4573-83df-8facd0dceed7">

And of course no style on that obfuscated class.

eg: `.nu7dy1l` which is `.note-title` and has a style of `letter-spacing: 1rem`, but there no style applied:

![image](https://github.com/soranoo/next-css-obfuscator/assets/147973044/ac90699a-e5c1-4fdf-8221-f44c0babe091)

![image](https://github.com/soranoo/next-css-obfuscator/assets/147973044/3b160a9b-3fdd-4394-9531-40c1aeb8baf7)

In expected:

![image](https://github.com/soranoo/next-css-obfuscator/assets/147973044/ef8d7e4e-f2f4-456f-97bb-c4571ecf284c)

**Comment by hoangnhan2ka3** [comment]
It works @soranoo 🚀
